text/0000-rand-core.md

```diff
@@ -0,0 +1,1099 @@
+- Feature Name: rand-core crate
+- Start Date: 2017-09-07
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Publish a new `rand-core` crate, containing:
+
+*   the `Rng` trait,
+*   possibly a `CryptoRng` trait,
+*   extension traits `SeedFromRng` and `SeedableRng`
+*   helper functions and/or default implementations for `Rng` functions,
+
+*For now*, re-export all the above in the `rand` crate. Also add two things to
+`rand`:
+
+*   the `NewSeeded` trait and its implementation for `SeedFromRng`
+*   a `Sample` trait
+
+## Links
+
+* [Sample implementation](https://github.com/dhardy/rand/tree/master/rand_core)
+* [Sample rustdoc](https://docs.rs/rand_core)
+* [RFC comments](https://github.com/rust-lang/rfcs/pull/2152)
+* [Rand crate revision RFC] (parent RFC)
+
+## Terminology
+
+For clarity, this document defines the following terms:
+
+*   **RNG**: Random Number Generator, a source of "randomness"
+*   **PRNG**: Pseudo-Random Number Generator (thus algorithmic and usually reproducible)
+*   **CSPRNG**: Cryptographically Secure PRNG
+
+## What this RFC is, and what it is not
+
+The [Rand crate revision RFC] has yielded a lot of useful suggestions, tweaks
+and refactors, but has proved to cover too much material to cover in a single
+thread. Despite many suggestions, it has also failed to reach consensus on a
+design for the core traits.
+
+This RFC covers:
+
+*   What core traits `rand` should expose
+*   How they should be published (new `rand-core` crate)
+*   How PRNG implementations should reference (crate dependencies) and implement
+    these traits
+*   Construction & seeding of PRNGs
+
+This RFC does not cover very much of the functionality in `rand`. In fact, the
+only aspects of this RFC which should affect most users (excluding RNG
+implementors and a few crypto users) is construction of PRNGs and the `Sample`
+extension trait (a new one; see below). Specifically, this RFC does not cover:
+
+*   How end-users should use `rand` (for now, the assumption is that all
+    necessary items will be re-exported through `rand` to minimise breakage;
+    this may change but will be the subject of a follow-up RFC, not this one)
```

Comment:
I would suggest deciding in *this* RFC whether we’re interested in keeping the reexport or would rather make the reexports deprecated somehow. There’s two sides to this. First, if we are interested in deprecation, it should land together with the split so hopefully we don’t have to go through the big breaking churn twice. Second, this would draw the guideline for the external RNG implementations. If those do “stuff” wrong, without being aware of our intentions for the rand/rand-core/etc, it will make churn ever so slightly worse.

I personally think re-exporting rand-core traits from `rand` (and crates implementing RNGs in general) is a significant usability boost – you wouldn’t need to `extern crate` two crates and be able to do `extern crate nicerng::{NiceRng, Rng}` for any RNG implementor you happen to use.

Reply:
Good point. I agree with you, pretty well, [and created this issue for discussion](https://github.com/dhardy/rand/issues/15).